clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3899,30 +3921,30 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
   if (KernelObj->isInvalidDecl())
     return;
 
-  // Calculate both names, since Integration headers need both.
-  std::string CalculatedName, StableName;
-  std::tie(CalculatedName, StableName) =
-      constructKernelName(*this, KernelCallerFunc, MC);
-  StringRef KernelName(IsSYCLUnnamedKernel(*this, KernelCallerFunc)
-                           ? StableName
-                           : CalculatedName);
+  {
+    // Do enough to calculate the StableName for the purposes of the hackery
+    // below for __pf_kernel_wrapper. Placed in a scope so that we don't
+    // accidentially use these values below, before the names are stabililzed.
+    std::string CalculatedName, StableName;
+    std::tie(CalculatedName, StableName) =
+        constructKernelName(*this, KernelCallerFunc, MC);
 
-  // Attributes of a user-written SYCL kernel must be copied to the internally
-  // generated alternative kernel, identified by a known string in its name.
-  if (StableName.find("__pf_kernel_wrapper") != std::string::npos)
-    copySYCLKernelAttrs(KernelObj);
+    // Attributes of a user-written SYCL kernel must be copied to the internally
+    // generated alternative kernel, identified by a known string in its name.
+    if (StableName.find("__pf_kernel_wrapper") != std::string::npos)
+      copySYCLKernelAttrs(KernelObj);
+  }
 
   bool IsSIMDKernel = isESIMDKernelType(KernelObj);
 
-  SyclKernelDeclCreator kernel_decl(*this, KernelName, KernelObj->getLocation(),
-                                    KernelCallerFunc->isInlined(),
-                                    IsSIMDKernel);
+  SyclKernelDeclCreator kernel_decl(*this, KernelObj->getLocation(),
+                                    KernelCallerFunc->isInlined(), IsSIMDKernel,
+                                    KernelCallerFunc);
```

Comment:
Yes.  I TRIED To use "SYCLKernel" to try to make it unambiguous, since it is the function marked with sycl_kernel.

Reply:
Yea I get it. Its confusing though, because we call the same function various names throughout the code - KernelCallerFunc, KernelFunc, SYCLKernel, etc. Maybe we should refactor this code (I don't mean in this patch) and rename it all to SYCLKernel to make the code easier to understand.

Reply:
I think that would be a valuable patch :)  The definition of 'kernel' in this project is pretty awful.  From my perspective, I consider that there are 2 types of kernels:  the sycl_kernel-instantiation, and the generated opencl kernel.  

At times in this code, we also seem to call the functor passed to the sycl-kernel-instantiation to ALSO be a kernel, which is horrifying.

Reply:
I've filed an issue to not lose track of this. I'll work on it in between tasks - https://github.com/intel/llvm/issues/4309